README.md
# <img alt="@bull-board" src="https://raw.githubusercontent.com/felixmosh/bull-board/master/packages/ui/src/static/images/logo.svg" width="35px" /> @bull-board

<p>
  <a href="https://github.com/tiagoboeing/bull-board/blob/master/LICENSE">
    <img alt="licence" src="https://img.shields.io/github/license/tiagoboeing/bull-board">
  </a>
  <img alt="open issues" src="https://img.shields.io/github/issues/tiagoboeing/bull-board"/>
<p>

**Docker version for `bull-board`.**

> Bull Dashboard is a UI built on top of [Bull](https://github.com/OptimalBits/bull) or [BullMQ](https://github.com/taskforcesh/bullmq) to help you visualize your queues and their jobs.
> With this library you get a beautiful UI for visualizing what's happening with each job in your queues, their status and some actions that will enable you to get the job done.

![UI](https://raw.githubusercontent.com/felixmosh/bull-board/master/screenshots/dashboard.png)

## Notes

This repository is a enhancement for original source at the repository [`felixmosh/bull-board`](https://github.com/felixmosh/bull-board)

**As this library provides only the visualization for your queues, keep in mind that:**

- You must have either [Bull](https://github.com/OptimalBits/bull) or [BullMQ](https://github.com/taskforcesh/bullmq) installed in your projects;
- Aside the options to retry and clean jobs, this library is not responsible for processing the jobs, reporting progress or any other thing. This must be done in your application with your own logic;
- If you want to understand the possibilities you have with the queues please refer to [Bull's docs](https://optimalbits.github.io/bull/) or [BullMQ's docs](https://docs.bullmq.io/);
- This library doesn't hijack Bull's way of working.

If you want to learn more about queues ([Bull](https://github.com/OptimalBits/bull) or [BullMQ](https://github.com/taskforcesh/bullmq)) and [Redis](https://redis.io/).

## Starting

### With Docker Compose

Declare the following service on your Docker Compose:

```docker
version: '3'

name: my-service
services:
  bull-monitor:
    container_name: bull-monitor
    image: tiagoboeing/bull-board:latest
    restart: always
    ports:
      - 4000:4000
    environment:
      QUEUE_PREFIX: 'bull'
      QUEUE_NAMES: |
        ExampleBullMQ;
        ExampleBull
    networks:
      - app-network
    depends_on:
      - redis

  redis:
    container_name: redis
    image: redis:6.2.3-alpine
    ports:
      - 6379:6379
    networks:
      - app-network

networks:
  app-network:
```

### Without Docker Compose

```docker
docker run \
  -e "QUEUE_PREFIX=bull" \
  -e "QUEUE_NAMES=ExampleBullMQ;ExampleBull" \
  tiagoboeing/bull-board:latest
```

> Check if Redis is Running and exposing ports!

## Available variables

See this `const` with [all available environments](https://github.com/tiagoboeing/bull-board/blob/1895a9c0b201d56d17ae2864cfcc0810a9b25dd3/src/index.ts#L15).

## Authentication

You can enable a login page to protect your dashboard with a simple username and password.

> This is a simple authentication, so take additional security measures if you need, something like restrict access to the dashboard only to internal IPs.

**Declare the following environment variables:**

```bash
AUTH_REQUIRE=true
AUTH_LOGIN=admin
AUTH_PASSWORD=admin
```

> Default login is `bull` and password is `board`. By default authentication is disabled.

## Developing

This project requires that you have a Redis instance running (you can simply run the `docker-compose.yml` to start the stack, use `npm run start:docker`).

Now, to try it out locally you can run:

```sh
npm run start:dev
```

The UI will be available at `http://localhost:4000`

# License

The original project is licensed under [MIT License](https://github.com/felixmosh/bull-board/blob/master/LICENSE), so it means it's completely free to use and copy. I decided to keep this available for all and free keeping the same license as the original on this repository and respecting the original creators.


CHANGELOG.md


### [1.2.1](https://github.com/tiagoboeing/bull-board/compare/v1.2.0-0...v1.2.1) (2023-08-15)


### Bug Fixes

* :adhesive_bandage: Fixed redirect to login page when using `basePath` env ([#5](https://github.com/tiagoboeing/bull-board/issues/5)) ([699b858](https://github.com/tiagoboeing/bull-board/commit/699b858a8d615398a213dd2df4bc2315484a400e))

## [1.2.0](https://github.com/tiagoboeing/bull-board/compare/v1.2.0-0...v1.2.0) (2023-06-29)

## [1.2.0-0](https://github.com/tiagoboeing/bull-board/compare/v1.1.0-0...v1.2.0-0) (2023-06-29)


### Features

* :sparkles: Created environment to enable basic auth using login and password ([#3](https://github.com/tiagoboeing/bull-board/issues/3)) ([0b994f7](https://github.com/tiagoboeing/bull-board/commit/0b994f7fc5a78fccca6af30ce97faf0f0b15cf9b))
* :stethoscope: Simple `/healthcheck` endpoint to be used on Kubernetes probes, like `liveness/readiness` ([#4](https://github.com/tiagoboeing/bull-board/issues/4)) ([3cda380](https://github.com/tiagoboeing/bull-board/commit/3cda380d17c2cd416934eef5aa83f7b15d4c2ef2))

## [1.1.0-0](https://github.com/tiagoboeing/bull-board/compare/v1.0.3-0...v1.1.0-0) (2023-04-27)


### Features

* :sparkles: Created environment variable `BASE_PATH` to work when using subpaths ([#2](https://github.com/tiagoboeing/bull-board/issues/2)) ([c236a53](https://github.com/tiagoboeing/bull-board/commit/c236a53011d69f5a9eaaea4d824c191d5ea9422e))

### [1.0.3-0](https://github.com/tiagoboeing/bull-board/compare/v1.0.2...v1.0.3-0) (2022-12-20)


### Others

* :construction_worker: Allow to generate pre-releases on CI ([71aa792](https://github.com/tiagoboeing/bull-board/commit/71aa792537c7510c3127c9115260c5720dc4c0aa))

### [1.0.2](https://github.com/tiagoboeing/bull-board/compare/v1.0.1...v1.0.2) (2022-12-20)


### Others

* :construction_worker: Build workflow now runs tests and build too with Node ([76106a3](https://github.com/tiagoboeing/bull-board/commit/76106a30b60c768d36c4b878e4877cc27c94fba1))
* :construction_worker: Updated Node version from `16` to `18` on GitHub release action ([6def9d3](https://github.com/tiagoboeing/bull-board/commit/6def9d32405bc8e48c580197dd23ace6686e79ec))

### [1.0.1](https://github.com/tiagoboeing/bull-board/compare/v1.0.0...v1.0.1) (2022-12-20)


### Bug Fixes

* :construction_worker: Fixed name of build workflow ([6195e7e](https://github.com/tiagoboeing/bull-board/commit/6195e7ea726a6ab263eb9fe31f384f2c12d01f03))


### Others

* :construction_worker: Created action to build code on push ([bd06b6f](https://github.com/tiagoboeing/bull-board/commit/bd06b6fe83442a9641fb08a003aa4254a8f9ea56))
* :memo: Breaking changes exclusive section on CHANGELOG.md ([afad36d](https://github.com/tiagoboeing/bull-board/commit/afad36d6c14c0dcbd8fe3b2e712d67880215081b))

## 1.0.0 (2022-12-20)


### Features

* :construction: Created template repository with Docker and environments ([3cd7af4](https://github.com/tiagoboeing/bull-board/commit/3cd7af4695966987a4c3fc8554a61f893f9429fb))


### Others

* :bug: Fixed `console.log` with UI link ([917b833](https://github.com/tiagoboeing/bull-board/commit/917b833b94ae9d699b66ca914ff87a0f2bc2bb17))
* :construction_worker: Created file to start release flow ([69ede41](https://github.com/tiagoboeing/bull-board/commit/69ede41904a04345bb6253d3147c463769c93aa5))
* :construction_worker: Created GitHub Action to build and push snapshots ([2f5fd58](https://github.com/tiagoboeing/bull-board/commit/2f5fd5879796b83c4dea2b47548ed05304c3c18f))
* :green_heart: Fixed name of action to create releases ([861baa7](https://github.com/tiagoboeing/bull-board/commit/861baa77863b88673607989a5c50c3a009d498fb))
* :green_heart: Fixed to CI use ref_name from GitHub context ([8c2a116](https://github.com/tiagoboeing/bull-board/commit/8c2a1167b10588e70d56cbef980789437fafefe7))
* :green_heart: Make repository checkout on GitHub Action ([75d0852](https://github.com/tiagoboeing/bull-board/commit/75d0852ee3a4017b06fa479bd1c931a9ef66e626))
* :green_heart: Replace branch name when contains `/` to `-` ([087db4e](https://github.com/tiagoboeing/bull-board/commit/087db4eb04edd23c4ca504688d02ef404787a8df))
* :package: Upgrade to Node version 18 ([0f2cebd](https://github.com/tiagoboeing/bull-board/commit/0f2cebd480495a35a25624e256f7d77d12a82814))
* :recycle: Move files to `src` folder and update Docker ([0d88763](https://github.com/tiagoboeing/bull-board/commit/0d88763c825e32ebc8b5cd967fc161e4f3be855b))
* :rotating_light: Fixes ESLint execution ([e5568da](https://github.com/tiagoboeing/bull-board/commit/e5568da6978a33f07b3167ffb3ca021c44d18409))
* :sparkles: Added Jest configs to enable testing ([26616de](https://github.com/tiagoboeing/bull-board/commit/26616de9aa19262e1d6e41c315a7404f4bd31f32))
* :white_check_mark: Created tests to `split-queue` ([21020b1](https://github.com/tiagoboeing/bull-board/commit/21020b1d0863c7b14df3f918a5a9e2036e1428d4))

src/utils/base-path/base-path.spec.ts
import { describe, test, expect } from '@jest/globals'
import { handleBasePath, handleRoutePath } from './base-path'

describe('BasePath', () => {
  test('should return "/" when path is undefined', () => {
    expect(handleBasePath()).toBe('/')
  })

  test('should return "/" when path is "/"', () => {
    expect(handleBasePath('/')).toBe('/')
  })

  test('should return path with "/" when passing without "/"', () => {
    expect(handleBasePath('test')).toBe('/test')
  })

  test('should return path with "/" when passing with "/"', () => {
    expect(handleBasePath('/test')).toBe('/test')
  })

  test('should work with subpaths', () => {
    expect(handleBasePath('test/subpath')).toBe('/test/subpath')
    expect(handleBasePath('/test/subpath')).toBe('/test/subpath')
  })
})

describe('handleRoutePath()', () => {
  test('should work without basePath', async () => {
    expect(handleRoutePath('login', '/')).toBe('/login')
    expect(handleRoutePath('/login', '/')).toBe('/login')

    expect(handleRoutePath('/login')).toBe('/login')
    expect(handleRoutePath('login')).toBe('/login')
    expect(handleRoutePath('login', '')).toBe('/login')
    expect(handleRoutePath('login', undefined)).toBe('/login')
    expect(handleRoutePath('login', null as never)).toBe('/login')
  })

  test('should work with basePath', async () => {
    expect(handleRoutePath('login', '/bull')).toBe('/bull/login')
    expect(handleRoutePath('/login', '/bull')).toBe('/bull/login')
  })
})


src/utils/split-queue-list/split-queue-list.spec.ts
import { describe, test, expect } from '@jest/globals'
import { splitQueueList } from './split-queue-list'

describe('SplitQueueList', () => {
  test('should works with only one queue', () => {
    const output = splitQueueList('queue1')
    expect(output).toEqual(['queue1'])
  })

  test('should works with repeated queue names', () => {
    const output = splitQueueList('queue1;queue1')
    expect(output).toEqual(['queue1', 'queue1'])
  })

  test('should split list without line breaks', () => {
    const output = splitQueueList('queue1;queue2')
    expect(output).toEqual(['queue1', 'queue2'])

    const output2 = splitQueueList('ExampleBullMQ;ExampleBull;')
    expect(output2).toEqual(['ExampleBullMQ', 'ExampleBull'])
  })

  test('should split list with line breaks', () => {
    const output = splitQueueList(`ExampleBullMQ;
ExampleBull`)
    expect(output).toEqual(['ExampleBullMQ', 'ExampleBull'])

    const output2 = splitQueueList(`ExampleBullMQ;
ExampleBull;`)
    expect(output2).toEqual(['ExampleBullMQ', 'ExampleBull'])
  })

  test('should trim empty chars', () => {
    const output = splitQueueList(' ExampleBullMQ; ExampleBull ')
    expect(output).toEqual(['ExampleBullMQ', 'ExampleBull'])

    const output2 = splitQueueList(' ExampleBullMQ; ExampleBull; ')
    expect(output2).toEqual(['ExampleBullMQ', 'ExampleBull'])
  })

  test('should trim empty chars with line breaks', () => {
    const output = splitQueueList(`ExampleBullMQ; 
    ExampleBull `)
    expect(output).toEqual(['ExampleBullMQ', 'ExampleBull'])

    const output2 = splitQueueList(`ExampleBullMQ; 
    ExampleBull; `)
    expect(output2).toEqual(['ExampleBullMQ', 'ExampleBull'])
  })

  describe('Special value', () => {
    test('should accept values with special types', () => {
      expect(() => splitQueueList('queue1-suffix')).not.toThrow()
      expect(() => splitQueueList('queue1-suffix;')).not.toThrow()
      expect(() => splitQueueList('queue1-suffix;queue2-suffix')).not.toThrow()

      expect(() => splitQueueList('queue1_suffix')).not.toThrow()
      expect(() => splitQueueList('queue1_suffix;')).not.toThrow()
      expect(() => splitQueueList('queue1_suffix;queue2_suffix')).not.toThrow()
    })
  })

  describe('Errors', () => {
    test('should throw error when queue list is empty', () => {
      const expectedError = /must be defined/

      expect(() => splitQueueList('')).toThrowError(expectedError)
      expect(() => splitQueueList('queue1,queue2')).toThrowError(expectedError)
      expect(() => splitQueueList('queue1|queue2')).toThrowError(expectedError)
      expect(() => splitQueueList('queue1/queue2')).toThrowError(expectedError)
      expect(() => splitQueueList('queue1\\queue2')).toThrowError(expectedError)
    })
  })
})


jest.config.js
/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/private/var/folders/1p/2bx3ymhj06780ln6g3kjt1_r0000gn/T/jest_dx",

  // Automatically clear mock calls, instances, contexts and results before every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  coverageDirectory: 'coverage',

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: 'v8',

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // The default configuration for fake timers
  // fakeTimers: {
  //   "enableGlobally": false
  // },

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "mjs",
  //   "cjs",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "json",
  //   "node"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest',

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state before every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state and implementation before every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // The number of seconds after which a test is considered as slow and reported as such in the results.
  // slowTestThreshold: 5,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  // testEnvironment: "jest-environment-node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.[jt]s?(x)",
  //   "**/?(*.)+(spec|test).[tj]s?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jest-circus/runner",

  // A map from regular expressions to paths to transformers
  transform: {
    '\\.js$': '<rootDir>/node_modules/babel-jest'
  }

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/",
  //   "\\.pnp\\.[^\\/]+$"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
}


src/index.ts
import { type BaseAdapter } from '@bull-board/api/dist/src/queueAdapters/base'
import {
  BullAdapter,
  BullMQAdapter,
  createBullBoard,
  ExpressAdapter
} from '@bull-board/express'
import styles from 'ansis'
import bodyParser from 'body-parser'
import Queue3 from 'bull'
import { type ConnectionOptions, Queue as QueueMQ } from 'bullmq'
import { ensureLoggedIn } from 'connect-ensure-login'
import express, {
  type NextFunction,
  type Request,
  type Response
} from 'express'
import session from 'express-session'
import morgan from 'morgan'
import passport from 'passport'
import { Strategy } from 'passport-local'
import path from 'path'
import { handleBasePath, handleRoutePath } from './utils/base-path/base-path'
import { splitQueueList } from './utils/split-queue-list/split-queue-list'

const environments = {
  port: process.env.PORT || 4000,
  queuePrefix: process.env.QUEUE_PREFIX,
  queueNames: process.env.QUEUE_NAMES || '',
  redisPort: parseInt(process.env.REDIS_PORT || '6379'),
  redisHost: process.env.REDIS_HOST || 'localhost',
  redisUsername: process.env.REDIS_USERNAME || 'default',
  redisPassword: process.env.REDIS_PASSWORD || '',
  basePath: process.env.BASE_PATH || '/',

  // auth
  authRequire: process.env.AUTH_REQUIRE === 'true',
  authLogin: process.env.AUTH_LOGIN || 'bull',
  authPassword: process.env.AUTH_PASSWORD || 'board'
}

const redisOptions: ConnectionOptions = {
  host: environments.redisHost,
  port: environments.redisPort,
  username: environments.redisUsername,
  password: environments.redisPassword
}

const createQueue3 = (name: string): Queue3.Queue<any> =>
  new Queue3(name, {
    redis: redisOptions,
    prefix: environments.queuePrefix
  })

const createQueueMQ = (name: string): QueueMQ<any, any, string> =>
  new QueueMQ(name, {
    connection: redisOptions,
    prefix: environments.queuePrefix
  })

const authMiddleware =
  () => (req: Request, _res: Response, next: NextFunction) => {
    if (!environments.authRequire || req.path.endsWith('/healthcheck')) {
      next()
      return
    }

    ensureLoggedIn({
      redirectTo: handleRoutePath('/login', environments.basePath)
    })(req, _res, next)
  }

const run = async (): Promise<void> => {
  console.log(styles.yellow('Starting Bull Board with: \n'))

  console.log('Redis')
  console.log(`- redis://${environments.redisHost}:${environments.redisPort}`)
  console.log(`- Username: ${environments.redisUsername}`)
  console.log(
    `- Password: ${environments.redisPassword ? '******' : '<empty>'}`
  )

  console.log('Configs')
  console.log(`- Auth required?: ${String(environments.authRequire)}`)
  console.log(`- Api base path: ${environments.basePath}`)
  console.log(`- Prefix: ${String(environments.queuePrefix)}`)
  console.log(`- Queues: ${environments.queueNames}`)
  console.log('')

  const app = express()

  const queueBullList: Array<Queue3.Queue<any>> = []
  const queueBullMqList: Array<QueueMQ<any, any, string>> = []

  const queuesList = splitQueueList(environments.queueNames)

  for await (const queueName of queuesList) {
    console.log(`Creating processor to queue: "${queueName}"`)

    const queueBull = createQueue3(queueName)
    const queueBullMq = createQueueMQ(queueName)

    queueBullList.push(queueBull)
    queueBullMqList.push(queueBullMq)
  }

  app.use('/add', (req, res) => {
    const opts = req.query.opts || ({} as any)

    if (opts.delay) {
      opts.delay = +opts.delay * 1000 // delay must be a number
    }

    const exampleBull = createQueue3('ExampleBull')
    const exampleBullMq = createQueueMQ('ExampleBullMQ')

    exampleBull.add({ title: req.query.title }, opts)
    exampleBullMq.add('Add', { title: req.query.title }, opts)

    queueBullList.forEach(async (queue) => {
      await queue.add({ title: req.query.title }, opts)
    })
    queueBullMqList.forEach(async (queue) => {
      await queue.add('Add', { title: req.query.title }, opts)
    })

    res.json({
      ok: true
    })
  })

  const basePath = handleBasePath(environments.basePath)
  console.log(`Base path: ${basePath}`)

  const serverAdapter: any = new ExpressAdapter()
  serverAdapter.setBasePath(basePath)

  const adaptersList: BaseAdapter[] = []

  queueBullList.forEach((queue) => adaptersList.push(new BullAdapter(queue)))
  queueBullMqList.forEach((queue) =>
    adaptersList.push(new BullMQAdapter(queue))
  )

  createBullBoard({ queues: adaptersList, serverAdapter })

  console.log('')
  console.log('Mapped Queues')
  console.log('Bull adapter:')
  console.log(`${queueBullList.map((queue) => `- ${queue.name}`).join('\n')}`)
  console.log('BullMq adapters:')
  console.log(`${queueBullMqList.map((queue) => `- ${queue.name}`).join('\n')}`)
  console.log('')

  if (environments.authRequire) {
    passport.use(
      new Strategy(function (username, password, cb) {
        if (
          username === environments.authLogin &&
          password === environments.authPassword
        ) {
          cb(null, { user: environments.authLogin })
          return
        }

        cb(null, false)
      })
    )

    passport.serializeUser(function (user: Express.User, done) {
      done(null, user)
    })

    passport.deserializeUser(function (user: Express.User, done) {
      done(null, user)
    })

    // Configure view engine to render EJS templates.
    app.set('views', path.resolve(__dirname, './views'))
    app.set('view engine', 'ejs')

    app.use(
      session({ secret: 'keyboard cat', saveUninitialized: true, resave: true })
    )
    app.use(passport.initialize({}))
    app.use(passport.session())

    app.use(bodyParser.urlencoded({ extended: false }))

    app.get(handleRoutePath('/login', environments.basePath), (req, res) => {
      res.render('login', { invalid: req.query.invalid === 'true' })
    })

    app.post(
      handleRoutePath('/login', environments.basePath),
      passport.authenticate('local', {
        failureRedirect: `${handleRoutePath(
          '/login',
          environments.basePath
        )}?invalid=true`
      }),
      (_req, res) => {
        res.redirect(handleBasePath(environments.basePath))
      }
    )
  }

  app.use(
    handleBasePath(environments.basePath),
    morgan('short'),
    authMiddleware(),
    serverAdapter.getRouter()
  )

  app.use('/healthcheck', (_req, res) => {
    res.send({ status: 'healthy' })
  })

  app.listen(environments.port, () => {
    console.log(`Running on ${environments.port}...`)
    console.log(`For the UI, open http://localhost:${environments.port}`)
    console.log(`Make sure Redis is running on port ${environments.redisPort}`)
    console.log('')
    console.log('To populate the queue, run:')
    console.log(
      `  curl http://localhost:${environments.port}/add?title=Example`
    )
    console.log('To populate the queue with custom options (opts), run:')
    console.log(
      `  curl http://localhost:${environments.port}/add?title=Test&opts[delay]=10`
    )
  })
}

run().catch((e) => {
  console.error(e)
})
